float32 density
float32 is_ok
float32 is_mid
float32 is_bad